wgsl/index.bs

```diff
@@ -600,28 +642,334 @@ storage_class
+* Section [[#layout-constraints]] describes the contraints on array strides and structure member
+    offsets, ultimately yielding definitions for
+    [=uniform buffer layout=] and [=storage buffer layout=].
+
+#### Memory Layout Intent #### {#memory-layout-intent}
+
+This section is informative, not normative.
+
+The layout rules describe two sets of constraints, one for uniform buffers and one for storage buffers.
+They are similar in many respects, but the uniform buffer layout is more restrictive.
+
+In particular:
+* Scalar data values are aligned to their own size.
+* There is no padding between components of a vector.
+* Two-element and four-element vectors are aligned to their size.
+* Three-element vectors are aligned as if they were four-element vectors.
+* An array's element alignment is a multiple of the element type's alignment.
+    * In a uniform buffer, the array element alignment is also a multiple of 16.
+* An array's alignment is the same as its element alignment.
+* A matrix with |N| columns is aligned as if it were an array of |N| column vectors.
+* A structure inherits the worst-case alignment of any of its members.
+    * In a uniform buffer, structure member alignment is also a multiple of 16.
+* The members of a structure are laid out in order: earlier members are appear earlier in
+    the buffer
+
+Additionally we define a value's allocation extent, or memory footprint, which determines
+how many memory locations must be reserved to store that value in host-shareable memory.
+Allocation extent is a determining factor of the minimum size of a buffer that can be bound
+to a uniform buffer variable
+or to a storage buffer variable.
+See [[#resource-layout-compatibility]].
+
+Compared to OpenGL:
+* OpenGL `std140` layout satisfies [SHORTNAME] [=uniform buffer layout=].
+* OpenGL `std430` layout satisifes [SHORTNAME] [=storage buffer layout=].
+* A type |S| with tightest offset and stride assignments satisfying [SHORTNAME] [=uniform buffer layout=] also satisfies OpenGL `std140` layout.
+* A type |S| with tightest offset and stride assignments satisfying [SHORTNAME] [=storage buffer layout=] also satisfies OpenGL `std430` layout.
+* OpenGL supports row-major matrices, but [SHORTNAME] does not.
+
+Compared to [[Vulkan1.2ext|Vulkan &sect; 15.6.4 Offset and Stride Assignment]]:
+* Vulkan *standard buffer layout* maps to [SHORTNAME] [=standard buffer layout rules=] with the following qualifications:
+    * Vulkan allows a vector to be aligned to the size of its scalar component, but [SHORTNAME] requires a more constrained alignment.
+    * The Vulkan `scalarBlockLayout` and `uniformBufferStandardLayout` features do not apply to [SHORTNAME].
+    * The Vulkan concept of *scalar alignment* does not correspond to a concept in [SHORTNAME];
+* The Vulkan *base alignment* for a type |S| corresponds to the [SHORTNAME] alignment requirement for |S| in the `storage` storage class: |Align|(|S|,`storage`).
```

Comment:
isn't it the other way around? I.e. "base alignment" == "uniform", and "extended alignment" == "storage"?

Reply:
You'd think it would be that way.  But it's not.  See [Offset and Stride Alignment](https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#interfaces-resources-layout)

> The *extended alignment* of the type of an OpTypeStruct member is similarly defined as follows:

...

> An array or structure type has an extended alignment equal to the largest extended alignment of any of its members, rounded up to a multiple of 16.

To me, that's one of the confusing things about how it's expressed in Vulkan. Especially because you need an *extension* to use the *base* alignments on uniform buffer, but you *don't* use an extension to use the *extended* aligments in a uniform buffer.  (Ok, I purposely put the most confusing aspects together here, but the point holds)

Reply:
I see, was reading that part wrong. Thanks for clarification!